### PR TITLE
[MLv2] Implement `zoom-in.binning` drill thru

### LINKED ISSUE #33575: [MLv2] [BE] Need num-bins and bin-width zoom drill thrus
There's some sort of zoom-in drill thru for binned columns, see

https://github.com/metabase/metabase/blob/0b0edf1b916aa6a4000767ec42a41c2721d22391/frontend/src/metabase-lib/queries/utils/drilldown.js#L364-L366

We need to port this over to MLv2. See https://github.com/metabase/metabase/issues/33556 and https://github.com/metabase/metabase/pull/33574 for an example of how we did the same thing for timeseries zooms.

### PR DESCRIPTION
**Pull request by camsaul**
Resolves #33575

Description copied from namespace docstring:

> This covers two types of 'zoom in' drills:
> 
> 1. If we have a query with a breakout with binning using the `:num-bins` strategy, return a drill that when applied
>    adds a filter for the selected bin ('zooms in') and changes the binning strategy to default. E.g.
> 
>        ORDERS + count aggregation + breakout on TOTAL (10 bins)
> 
>        =>
> 
>        Click the 40-60 bin in the results (returned by the QP as `40`) and choose 'Zoom In'
> 
>        =>
> 
>        ORDERS + count aggregation + filter TOTAL between 40 and 60 + breakout on TOTAL (auto bin)
> 
>    Note that we need to look at the fingerprint info in the column metadata to determine how big each bin
>    is (e.g. to determine each bin was 20 wide) -- this uses [[lib.binning.util/nicer-bin-width]], which is what the QP
>    uses.
> 
>    In other words, this bin adds a filter for the selected bin and the replaces the breakout binning with a `:default`
>    binning strategy.
> 
> 2. Breakout with binning with `:bin-width`:
> 
>        PEOPLE + count aggregation + breakout on LATITUDE (bin width: 1°)
> 
>        =>
> 
>        Click on the 41°-42° bin in the results (returned by the QP as `41`) and choose 'Zoom In'
> 
>        =>
> 
>        PEOPLE + count aggregation + filter LATITUDE between 41 and 42 + breakout on LATITUDE (bin width: 0.1°)
> 
>    In other words, this bin adds a filter for the selected bin and then divides the bin width in the breakout binning
>    options by 10.

As part of this I moved some of the code in `metabase.query-processor.middleware.binning` that calculates what bin widths will be used for `:num-bins` to Cljc. Presumably the frontend had to actually run the query and calculate this from the results or QP metadata somehow (not sure how it figured this out), but with MLv2 we can do one better and just use the same code the QP uses to calculate the bin size.

As part of this I had to fix `metabase.util/round-to-decimals` for JS, it was broken -- it used `Number.precision()`, which rounds it to a precision including digits to the left of the decimal place, e.g. `(u/round-to-decimals 2 12345.6789) => 12000.0` when what we wanted was `12345.67`. That is now fixed

### CODE REVIEW

File: test/metabase/lib/binning/util_test.cljc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+(deftest ^:parallel ceil-to-test
+  (are [precision x expected] (= expected
+                                 (#'lib.binning.util/ceil-to precision x))
+    1.0  1    1.0
+    1.0  1.1  2.0
+    1.0  1.8  2.0
+    15.0 1.0  15.0
+    15.0 15.0 15.0
+    15.0 16.0 30.0))
```
Comment (metamben) [inline test/metabase/lib/binning/util_test.cljc:22]: Are negative numbers intentionally omitted?

**Review by metamben — APPROVED** [review]
LGTM

File: src/metabase/lib/binning/util.cljc [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            (let [candidate-width (* pleasing-number scale)]
+              (when (>= candidate-width min-bin-width)
+                candidate-width)))
+          pleasing-numbers)))
+
+(mu/defn ^:private nicer-bounds :- [:tuple number? number?]
+  [min-value :- number?
+   max-value :- number?
+   bin-width :- ::lib.schema.binning/bin-width]
+  [(floor-to bin-width min-value) (ceil-to bin-width max-value)])
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/binning/util.cljc:89]: `::lib.schema.binning/bin-width` is an alias for `::lib.schema.common/positive-number`. It doesn't make sense to have a bin-width of zero anyway

File: test/metabase/lib/binning/util_test.cljc (hunk 1 above)
Comment (camsaul) [reply test/metabase/lib/binning/util_test.cljc:22]: Nope. I'll add tests

File: src/metabase/lib/binning/util.cljc [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+(defn- ceil-to
+  [precision x]
+  (let [scale (/ precision)]
+    (/  (math/ceil (* x scale)) scale)))
+
+(defn- floor-to
+  [precision x]
+  (let [scale (/ precision)]
+    (/ (math/floor (* x scale)) scale)))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/binning/util.cljc:63]: i don't really know why this code was done this way, I just moved it. But it looks like you're right WRT to simplifying it so I'll update it.

File: src/metabase/lib/binning/util.cljc [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+   [metabase.util :as u]
+   [metabase.util.malli :as mu]))
+
+(mu/defn ^:private calculate-bin-width :- ::lib.schema.binning/bin-width
+  "Calculate bin width required to cover interval [`min-value`, `max-value`] with `num-bins`."
+  [min-value :- number?
+   max-value :- number?
+   num-bins  :- ::lib.schema.binning/num-bins]
+  (u/round-to-decimals 5 (/ (- max-value min-value)
+                            num-bins)))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/lib/binning/util.cljc:18]: bin width has to be a positive number

**Pull request merged**